PAR::Packer is not buildable on relocatable-perl 5.40.0.1

This is perhaps not directly related to relocatable-perl, but I'm sharing some brief finding here first.

Basically, `cpm install -g PAR::Packer` failed in relocatable-perl 5.40.0.1 with this compilation error:

```
cc main.o sha1.o  -s -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/home/gugod/perl5/perlbrew/perls/skaji-relocatable-perl-5.40.0.1/lib/5.40.0/x86_64-linux/CORE -lperl -lpthread -ldl -lm -lcrypt -lutil -lc  -o ./par
/usr/bin/ld: /home/gugod/perl5/perlbrew/perls/skaji-relocatable-perl-5.40.0.1/lib/5.40.0/x86_64-linux/CORE/libperl.a(perl.o): in function `perl_construct':
/tmp/perl-build-1728287368.1/perl-5.40.0/perl.c:326:(.text+0x281b): undefined reference to `_LIB_VERSION'
collect2: error: ld returned 1 exit status
```

Here's how I could reproduce it on my Fedora Linux:

```
# uname -srvmo  
Linux 6.11.5-300.fc41.x86_64 #1 SMP PREEMPT_DYNAMIC Tue Oct 22 20:11:15 UTC 2024 x86_64 GNU/Linux


# perlbrew --version
/home/gugod/perl5/perlbrew/bin/perlbrew  - App::perlbrew/1.01

## This step would download and install this tarball: https://github.com/skaji/relocatable-perl/releases/download/5.40.0.1/perl-linux-amd64.tar.gz
# perlbrew install skaji-relocatable-perl-5.40.0.1 

# perlbrew use skaji-relocatable-perl-5.40.0.1 

# which perl                                  
~/perl5/perlbrew/perls/skaji-relocatable-perl-5.40.0.1/bin/perl

# perl -V | grep config
Summary of my perl5 (revision 5 version 40 subversion 0) configuration:
    config_args='-Dprefix=/opt/perl -Duserelocatableinc -Dman1dir=none -Dman3dir=none -DDEBUGGING=-g -des -Ui_xlocale -Dlibs=-lpthread -ldl -lm -lcrypt -lutil -lc -Dd_crypt -Dlibpth=/lib /lib/x86_64-linux-gnu /lib64 /usr/lib /usr/lib/x86_64-linux-gnu /usr/lib64 /usr/local/lib /usr/local/lib64'

# cpm install -g --show-build-log-on-failure PAR::Packer
.......
2024-12-23T08:41:34,118330,PAR-Packer-1.063| cc main.o sha1.o  -s -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/home/gugod/perl5/perlbrew/perls/skaji-relocatable-perl-5.40.0.1/lib/5.40.0/x86_64-linux/CORE -lperl -lpthread -ldl -lm -lcrypt -lutil -lc  -o ./par
2024-12-23T08:41:34,118330,PAR-Packer-1.063| /usr/bin/ld: /home/gugod/perl5/perlbrew/perls/skaji-relocatable-perl-5.40.0.1/lib/5.40.0/x86_64-linux/CORE/libperl.a(perl.o): in function `perl_construct':
2024-12-23T08:41:34,118330,PAR-Packer-1.063| /tmp/perl-build-1728287368.1/perl-5.40.0/perl.c:326:(.text+0x281b): undefined reference to `_LIB_VERSION'
2024-12-23T08:41:34,118330,PAR-Packer-1.063| collect2: error: ld returned 1 exit status
2024-12-23T08:41:34,118330,PAR-Packer-1.063| make[1]: *** [Makefile:887: par] Error 1
2024-12-23T08:41:34,118330,PAR-Packer-1.063| make[1]: Leaving directory '/home/gugod/.perl-cpm/work/1734910891.118325/PAR-Packer-1.063/myldr'
2024-12-23T08:41:34,118330,PAR-Packer-1.063| make: *** [Makefile:547: subdirs] Error 2
...

* The issue `undefined reference to _LIB_VERSION` will be fixed by compiling relocatabale-perls with `-Ud_libm_lib_version`.
* However, even if we build relocatabale-perls with `-Ud_libm_lib_version`, creating executables with libperl.a would fail with `relocation R_X86_64_32 against symbol PL_memory_wrap can not be used when making a PIE object; recompile with -fPIE`. This is because gcc in recent OSes configured with `--enable-default-pie`.

--- a/build/Dockerfile
+++ b/build/Dockerfile
@@ -1,35 +1,49 @@
-FROM buildpack-deps:bookworm AS static-libcrypt
+FROM ubuntu:18.04 AS static-libcrypt
 
 RUN set -eux; \
-  mkdir /libc6-dev; \
-  cd /tmp; \
-  if [ $(uname -m) = x86_64 ]; then \
-    curl -fsSL -O http://security.ubuntu.com/ubuntu/pool/main/g/glibc/libc6-dev_2.23-0ubuntu3_amd64.deb; \
-    dpkg-deb -x libc6-dev_2.23-0ubuntu3_amd64.deb /libc6-dev; \
-    cp /libc6-dev/usr/lib/x86_64-linux-gnu/libcrypt.a /libcrypt.a; \
-  else \
-    curl -fsSL -O http://ports.ubuntu.com/pool/main/g/glibc/libc6-dev_2.23-0ubuntu3_arm64.deb; \
-    dpkg-deb -x libc6-dev_2.23-0ubuntu3_arm64.deb /libc6-dev; \
-    cp /libc6-dev/usr/lib/aarch64-linux-gnu/libcrypt.a /libcrypt.a; \
-  fi; \
-  :
-
-FROM centos:centos7 AS builder
-
-# see https://gist.github.com/skaji/76203327b517cb44da88a4301de118d3
-RUN sed -i 's/override_install_langs=en_US.UTF-8/override_install_langs=en_US.utf8/' /etc/yum.conf
-
-# https://stackoverflow.com/questions/78692851/could-not-retrieve-mirrorlist-http-mirrorlist-centos-org-release-7arch-x86-6
-RUN sed -i 's/mirrorlist/#mirrorlist/g' /etc/yum.repos.d/CentOS-*
-RUN sed -i 's|#baseurl=http://mirror.centos.org|baseurl=https://vault.centos.org|g' /etc/yum.repos.d/CentOS-*
-
-RUN yum install -y \
+  export DEBIAN_FRONTEND=noninteractive; \
+  apt-get update; \
+  apt-get install -y \
+    ca-certificates \
+    curl \
     gcc \
+    libc6-dev \
     make \
+    patch \
+    perl \
     tar \
     wget \
+    xz-utils \
+    ;
+
+ARG LIBXCRYPT_VERSION=4.4.36
+
+RUN set -eux; \
+  cd /tmp; \
+  curl -fsSL -o libxcrypt-${LIBXCRYPT_VERSION}.tar.xz https://github.com/besser82/libxcrypt/releases/download/v${LIBXCRYPT_VERSION}/libxcrypt-${LIBXCRYPT_VERSION}.tar.xz; \
+  tar xf libxcrypt-${LIBXCRYPT_VERSION}.tar.xz; \
+  cd libxcrypt-${LIBXCRYPT_VERSION}; \
+  ./configure --with-pic --disable-shared; \
+  make; \
+  mv .libs/libcrypt.a /; \
+  :
+
+# XXX We use ubuntu:18.04 because its gcc is configured with --enable-default-pie
+FROM ubuntu:18.04 AS builder
+
+RUN set -eux; \
+  export DEBIAN_FRONTEND=noninteractive; \
+  apt-get update; \
+  apt-get install -y --no-install-recommends \
+    ca-certificates \
+    gcc \
+    libc6-dev \
+    make \
     patch \
-    xz
+    tar \
+    wget \
+    xz-utils \
+    ;
 RUN mkdir -p \
     /lib \
     /lib/$(uname -m)-linux-gnu \
@@ -38,25 +52,25 @@ RUN mkdir -p \
     /usr/lib/$(uname -m)-linux-gnu \
     /usr/lib64 \
     /usr/local/lib \
-    /usr/local/lib64
+    /usr/local/lib64 \
+    ;
 RUN wget -q -O - https://raw.githubusercontent.com/skaji/relocatable-perl/main/perl-install | bash -s /perl
 RUN wget -q -O /cpm https://raw.githubusercontent.com/skaji/cpm/main/cpm
 RUN --mount=type=bind,target=src /perl/bin/perl /cpm install -g --cpmfile src/build/cpm.yml
 
-RUN rm -f /usr/lib64/libcrypt.so /usr/lib64/libcrypt.so.1
-COPY --from=static-libcrypt /libcrypt.a /usr/lib64/libcrypt.a
+RUN rm -f /usr/lib/$(uname -m)-linux-gnu/libcrypt.so
+COPY --from=static-libcrypt /libcrypt.a  /tmp/libcrypt.a
+RUN cp -f /tmp/libcrypt.a /usr/lib/$(uname -m)-linux-gnu/libcrypt.a
+
 RUN --mount=type=bind,target=src /perl/bin/perl src/build/relocatable-perl-build --perl_version $(cat src/BUILD_VERSION) --prefix /opt/perl
-RUN cd /usr/lib64; \
-  rm -f libcrypt.a; \
-  ln -sf libcrypt-2.17.so libcrypt.so; \
-  ln -sf libcrypt-2.17.so libcrypt.so.1; \
-  :
+
+RUN ln -sf /lib/$(uname -m)-linux-gnu/libcrypt.so.1 /usr/lib/$(uname -m)-linux-gnu/libcrypt.so
 
 RUN /opt/perl/bin/perl /cpm install -g App::cpanminus App::ChangeShebang
 RUN /opt/perl/bin/change-shebang -f /opt/perl/bin/*
 RUN set -eux; \
   cd /tmp; \
-  _ARCHNAME=$(if [[ $(uname -m) = x86_64 ]]; then echo amd64; else echo arm64; fi); \
+  _ARCHNAME=$(if [ $(uname -m) = x86_64 ]; then echo amd64; else echo arm64; fi); \
   cp -r /opt/perl perl-linux-$_ARCHNAME; \
   tar cf perl-linux-$_ARCHNAME.tar perl-linux-$_ARCHNAME; \
   gzip -9 --stdout perl-linux-$_ARCHNAME.tar > /perl-linux-$_ARCHNAME.tar.gz; \

--- a/build/relocatable-perl-build
+++ b/build/relocatable-perl-build
@@ -22,7 +22,14 @@ sub run {
     my ($cmd, $log) = @_;
     warn "---> @$cmd\n";
     run3 $cmd, undef, $log, $log, { append_stdout => 1 };
-    $? == 0 or die "failed @$cmd";
+    if ($? != 0) {
+        open my $fh, "<", $log or die;
+        while (<$fh>) {
+            print;
+        }
+        die;
+    }
+    return 1;
 }
 
 =head1 NAME
@@ -138,6 +145,9 @@ sub perl_build {
         # RHEL9 removes libcrypt.so.1 by default; so we will link libcrypt.a statically
         # manually define d_crypt here
         push @Configure, "-Dd_crypt";
+
+        # math.h in debian does not define _LIB_VERSION
+        push @Configure, "-Ud_libm_lib_version";
 
         my $arch = (uname)[4];
         my @libpth = (

